### ISSUE TITLE
notEmpty attribute option doesn't work if MySql table exists.

### ISSUE THREAD

**Issue by skywickenden** [balderdashy/waterline#1047]
If a model attribute has a notEmpty option and the model reflects a mySql table that does exist then it fails to assert that the attribute is empty.

E.G.

Given this Sails model:

    module.exports = {
        tableName: 'foo',
        autoPK: false,
        autoCreatedAt: false,
        autoUpdatedAt: false,
        attributes: {
            id: {
                type: 'integer',
                unique: true,
                primaryKey: true,
                autoIncrement: true,
                columnName: 'id'
            },
            bar: {
                type: 'string',
                columnName: 'bar',
                size: 127,
                notEmpty: true
            }
        }
    };

And this test (uses 'mocha' and 'should')

    describe.only('FooModel', function() {
        describe('Try to insert an empty value', function() {
            it('Should not allow insertion of an empty string into the bar column.', function (done) {
                FooModel
                    .create({
                        bar: '',
                    })
                    .exec(function onCreated(err, created) {
                        (err === null).should.be.false;
                    done();
                    });
            });
        });
    });

This test will work as expected. An error will be thrown and the test will succeed.

Now create the table in MySQL that refers to the model:

	CREATE TABLE `foo` (
		`id` INT(10) NOT NULL AUTO_INCREMENT,
		`bar` VARCHAR(127) NOT NULL DEFAULT '0',
		PRIMARY KEY (`id`)
	)
	COLLATE='utf8_unicode_ci'
	ENGINE=InnoDB
	AUTO_INCREMENT=1;

Now rerun the test. The test fails to throw an error. (err is null). Causing the test to fail. A row of data is also created in the table.

**Comment by dmarcelino** [comment]
Hi @skywickenden, 

Validations in waterline are handled by `anchor`, for 'notEmpty' they are handled by the code in https://github.com/balderdashy/anchor/blob/0e7d553ae2dfdd2855910d6e07dc8b435fc681dc/lib/match/rules.js#L30-L38

`anchor` itself doesn't have a 'notEmpty' test, maybe we should start there.

**Comment by skywickenden** [comment]
@dmarcelino Thanks for lead. I'll have a look.

**Comment by skywickenden** [comment]
I've tracked the problem down to this section in [validations,js](https://github.com/balderdashy/waterline/blob/master/lib/waterline/core/validations.js#L132-L136)

    if(!curValidation.required) {
      if(value === null || value === '') return cb();
    }

Essentially: if 'required' is not set and the value is empty then it never even tests to see if it is notEmpty.

The definition of the various empty, null, required validators was decided in [this issue](https://github.com/balderdashy/waterline/issues/603). This gives the following meanings:

required: cannot be empty,null or undefined
empty: cannot have content, be null or undefined. Must be empty (does this imply it must be a string?)
null: Must be null
notNull: Must not be null
empty: Must be empty
notEmpty: Must not be empty
undefined: Must be undefined.

Quite a few of these will fail in some situations due to the code above. I've checked the tests and there are no tests for these validators except the required one. I'm happy to write a patch and some tests but as I'm new around here can the community confirm that I am understanding the meaning of the validators correctly before I start.

**Comment by dmarcelino** [comment]
I share the same opinion as @randallmeeker expressed in https://github.com/balderdashy/waterline/issues/603#issuecomment-53018751:
> There is ```null``` and ```notNull``` and ```empty``` and ```notEmpty``` and ```undefined``` validators. I think those should be used where you need a more fine grained approach like your use case. I think ```required``` is non-specific and an acceptable catch all for empty, null or undefined. Since required is more ambiguous it makes sense that way (or at least to me). 

So...

1. `required`: cannot be empty, null or undefined
2. `empty`: cannot have content: **must be empty string, array or object** (anything else that can be empty?)
3. `notEmpty`: Must not be empty **but it can be null or undefined**
4. `null`: Must be null
5. `notNull`: Must not be null
6. `undefined`: Must be undefined.
7. `notUndefined`: **does it make sense to add this?**

*I've highlighted the parts where my opinion differs than what you've wrote*

It's also important to note that `notNull` doubles as a DB constraint so in an SQL DB an attempt to save `undefined` would likely fail independently of the validation only being applied to `null`. This raises the question if we should distinguish `null` from `undefined`...

> I'm happy to write a patch and some tests

Yes, please! :grinning: 

@particlebanana, @devinivy, @tjwebb, @RWOverdijk what say you?

**Comment by skywickenden** [comment]
@dmarcelino Your comments all make good sense. (Can't believe I forgot about empty arrays and objects - shows my roots in SQL land).

**Comment by RWOverdijk** [comment]
@dmarcelino I agree with you. Though I do think we should distinguish null from undefined. They can both have a different meaning.

**Comment by skywickenden** [comment]
Two options RE notNull  and undefined.

We could have a notSqlNull and sqlNull, which would treat undefined as a null. The notNull could do exactly as it says. This would then require refactoring the use of notNull for the DB constraint into notSqlNull.  notNull could still be used as a DB constraint to prevent backwards compatibility issues.

Alternatively the SQL adapters could convert undefineds to null if the field is marked as notNull, but this could cause unexpected behavior and so is also not ideal.

**Comment by dmarcelino** [comment]
I don't really like `(not)SqlNull`... I think so far waterline has been able to avoid having options that are too specific to one DB type in the core code.

Adapters should handle it, if there is any need for it. Assuming we add `notUndefined` a user could use:
```js
        bar: {
            type: 'string',
            notNull: true,
            notUndefined: true
        }
```
If he/she want to be extra careful.

**Comment by Ryanc1256** [comment]
@dmarcelino Agreed, Core should be only really for code that handles things like models, validation and all that and it's up to the adapters to support something like "notNull" or any database specific code.

**Comment by skywickenden** [comment]
I agree the adapter should handle it. The adapter shouldn't just ignore it though - that could lead to opaque bugs. If I try to insert a value of undefined for a field that does not allow nulls MySql currently throws 

    Details:  Error: ER_BAD_NULL_ERROR: Column 'bar' cannot be null

This is not ideal.

If the field does allow nulls then Null is inserted instead of undefined. Also not good.

Perhaps this should be raised as a separate issue though as the initial bug here could be solved without this?

**Comment by dmarcelino** [comment]
Yes, that issue should be raised in the [sails-mysql project](https://github.com/balderdashy/sails-mysql).

**Comment by skywickenden** [comment]
Ok, I'll start work on this with the spec outlined by @dmarcelino

**Comment by Ryanc1256** [comment]
Cool :+1:

**Comment by skywickenden** [comment]
Should the number zero (0) be considered empty? (Currently the code considers it to be empty?)

**Comment by skywickenden** [comment]
@dmarcelino 

I've written the tests and have 11 failing at the moment.

Quite a few of them differ from rules.js as described by anchor. Should I be creating overrides in validations.js or be raising an issue in the anchor project and handling them there?

**Comment by dmarcelino** [comment]
Hi @skywickenden, I think the anchor project is the more natural place for these, thanks.

**Comment by dmarcelino** [comment]
Lets continue discussing this issue in balderdashy/waterline-docs#86.

**Comment by sailsbot** [comment]
Thanks for posting, @skywickenden.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
